### ISSUE TITLE
zoom to extent error for Postgis

### ISSUE THREAD

**Issue by amyleew** [mapbox/mapbox-studio-classic#1487]
Clicking "zoom to extent" button returns error: `Cannot access source metadata. Error: ENOENT, open 'undefined'` in source mode 

![image](https://cloud.githubusercontent.com/assets/4587826/10684734/e5ee54ee-791e-11e5-9528-0bad09b3efe5.png)

**Comment by wilhelmberg** [comment]
dupe of https://github.com/mapbox/mapbox-studio-classic/issues/995

---

@springmeyer not yet arrived at the real PostGIS problem, but came across this on the way.

https://github.com/mapbox/mapbox-studio-classic/commit/9a3b92a1157beb9a0bd2b83ef66b1062ed63d1af introduces

``` js
        function clamp(value,expected) {
            if (value > expected || value < expected) return expected;
            return value;
        }
        map.fitBounds([
            [clamp(extent[1],-180),clamp(extent[0],-85.0511)],
            [clamp(extent[3],180),clamp(extent[2],85.0511)]
        ],{'animate': false});
```

I suppose the intent is to get an extent that's always within the Web Mercator maximum bbox?
But 

``` js
(value > expected || value < expected)
```

is always `true`, except when `value` equals `expected` exactly, always returning the max bbox, even for spatially very small datasets.

---

Maybe this is a better implementation: 
- clips at WebMerc bounds if extent is too big 
- and also returns the correct extent for small areas:

``` js
        function clamp(value,min,max,retval) {
            if ((value < min) || (value > max)) return retval;
            return value;
        }
        map.fitBounds([
            [clamp(extent[1],-180,180,-180),clamp(extent[0],-85.0511,85.0511,-85.0511)],
            [clamp(extent[3],-180,180,180),clamp(extent[2],-85.0511,85.0511,85.0511)]
        ],{'animate': false});
```

Or should always the whole world bbox be returned?

**Comment by wilhelmberg** [comment]
@amyleew could you give this branch a spin (with PostGIS and file based data):
https://github.com/mapbox/mapbox-studio-classic/tree/src-zoom-to-extent

Should solve `zoom to extent` with:
- PostGIS based layers and also
- file based layers with small spatial extent

**Comment by amyleew** [comment]
@BergWerkGIS - Results for clicking `zoom to extent` on this branch:
- Works perfectly with file based layers with small spatial extent
- With PostGIS data, clicking it for each layer takes me to `-0.0000, -0.0000` `z22`:
  ![image](https://cloud.githubusercontent.com/assets/4587826/10732545/f7157838-7bd0-11e5-9f53-e230dca0a893.png)

**Comment by wilhelmberg** [comment]
@springmeyer 

tl;dr:
`node-mapnik`'s `ds.extent()` is broken with PostGIS layers when `scale_denominator` is used in the table's SQL query and `extent_from_subquery` is set to `true`.

---

#### problem

I used https://github.com/mapbox/mapnik-test-data/tree/master/data/shp/world_merc for testing.

[`lib/source.js::source->extent`](https://github.com/mapbox/mapbox-studio-classic/blob/281d51360dbe7b90187a32779c91bf1a53031962/lib/source.js#L299-L351) always returned the full extent of the PostGIS table, even for queries like this:

``` sql
(select *
from world_merc
where
  Iso2='AR'
  AND geom && !bbox!
) as data
```

So I added `Datasource.extent_from_subquery = true;` which worked fine for me and correctly zoomed to Argentina, but not for @amyleew getting extents like

```
[0.00000001, 0000001, 0.000000011, 00000011]
```

After comparing our queries I was able to reduce the problem to this:

``` sql
(select *
from world_merc
where
  Iso2='AR'
  AND z(!scale_denominator!) BETWEEN 0 AND 22
  AND geom && !bbox!
) as data
```

---

#### conclusion
- `ds.extent()` always returns the full extent of the original table
- `ds.extent()` + `extent_from_subquery=true` returns the extent of the query
- `ds.extent()` + `extent_from_subquery=true` + `scale_denominator`: **broken**
- `ds.extent()` + `extent_from_subquery=false` + `scale_denominator`: works, but always returns the full extent of the original table

---

#### current solution

My current solution is to check, if `scale_denominator` exists in the table's SQL query.
If it is, don't add `extent_from_subquery=true`, otherwise add.

https://github.com/mapbox/mapbox-studio-classic/blob/281d51360dbe7b90187a32779c91bf1a53031962/lib/server.js#L586-L590

``` js
            if (data.Layer[0].Datasource.table) {
                if (!/scale_denominator/i.test(data.Layer[0].Datasource.table)) {
                    data.Layer[0].Datasource.extent_from_subquery = true;
                }
            }
```

This seems a bit hackish, as it would be possible, that `scale_denominator` is present, but commented:

``` sql
(select *
from world_merc
where
  Iso2='AR'
  -- AND z(!scale_denominator!) BETWEEN 0 AND 22
  AND geom && !bbox!
) as data
```

Any ideas for a "proper" solution?

**Comment by amyleew** [comment]
@BergWerkGIS - Seems to be functioning as expected in UI and map view in the canvas. Zooming properly based on my data.

**Working with this example:**
![image](https://cloud.githubusercontent.com/assets/4587826/10761441/ca95c0d4-7c97-11e5-959c-4ed3c1786718.png)

**Returns:**

``` sh
server, req.query: { ds: 'marine_label' }
server, req.query.ds: marine_label
server, req.headers.referer: http://localhost:3001/source?id=tmsource:///Users/amyleewalton/amyleew/mapbox-streets-source.tm2source
server info, err: null
ds_extent: [ -19668274.0468755,
  -13350830.1840136,
  19796362.9106962,
  18147518.3456762 ]
projection: +proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0.0 +k=1.0 +units=m +nadgrids=@null +wktext +no_defs +over
current: {}
transformation
extent: [ -176.6831118855957,
  -75.94297067802322,
  177.83375372655215,
  83.34758186794012 ]
server info, extent: [ -176.6831118855957,
  -75.94297067802322,
  177.83375372655215,
  83.34758186794012 ]
```

**This example:**
![image](https://cloud.githubusercontent.com/assets/4587826/10761561/5339bd96-7c98-11e5-8a9f-299118c90769.png)

**Returns:**

``` sh
server, req.query: { ds: 'barrier_line' }
server, req.query.ds: barrier_line
server, req.headers.referer: http://localhost:3001/source?id=tmsource:///Users/amyleewalton/amyleew/mapbox-streets-source.tm2source
server info, err: null
ds_extent: [ -8983326.22, 4876206.97, -8828955.13, 4993859.48 ]
projection: +proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0.0 +k=1.0 +units=m +nadgrids=@null +wktext +no_defs +over
current: {}
transformation
extent: [ -80.69859245657646,
  40.07059986827245,
  -79.31185336084457,
  40.87457778315463 ]
server info, extent: [ -80.69859245657646,
  40.07059986827245,
  -79.31185336084457,
  40.87457778315463 ]
```

**Comment by wilhelmberg** [comment]
@amyleew Great, thanks for testing.

**Comment by springmeyer** [comment]
@amyleew - I'm unclear. Did not work before, now works for you. What changed?

**Comment by amyleew** [comment]
Yes @springmeyer. PostGIS `zoom to extent` button was poping up [an error yesterday](https://github.com/mapbox/mapbox-studio-classic/issues/1487#issuecomment-151169032) for PostGIS data using double queries for `z(!scale_denominator!)` and `geom && !bbox!`

@BergWerkGIS made a fix to [lib/server.js#L586-L590](https://github.com/mapbox/mapbox-studio-classic/blob/281d51360dbe7b90187a32779c91bf1a53031962/lib/server.js#L586-L590) but it may not the most elegant way? (I think that's what he's wondering...)
